url7f729337e3_timmolter_Sundial_tgz-pJ8-org_knowm_sundial_SampleRunJ8/src/org/knowm/sundial/SundialJobScheduler.java

```diff
@@ -0,0 +1,534 @@
+
+    addJob(jobName, jobClassName, null, false);
+
+  }
+
+  /**
+   * Adds a Job to the scheduler. Replaces a matching existing Job.
+   *
+   * @param jobName
+   * @param jobClass
+   * @param params Set this null if there are no params
+   * @param isConcurrencyAllowed
+   */
+  public static void addJob(String jobName, Class<? extends Job> jobClass, Map<String, Object> params, boolean isConcurrencyAllowed)
+      throws SundialSchedulerException {
+
+    try {
+      JobDataMap jobDataMap = new JobDataMap();
+      if (params != null) {
+        for (Entry<String, Object> entry : params.entrySet()) {
+          jobDataMap.put(entry.getKey(), entry.getValue());
+        }
+      }
+
+      JobDetail jobDetail = newJobBuilder(jobClass).withIdentity(jobName).usingJobData(jobDataMap).isConcurrencyAllowed(isConcurrencyAllowed).build();
+
+      getScheduler().addJob(jobDetail);
+
+    } catch (SchedulerException e) {
+      logger.error("ERROR ADDING JOB!!!", e);
+      throw new SundialSchedulerException("ERROR ADDING JOB!!!", e);
+    }
+  }
+
+  /**
+   * Adds a Job to the scheduler. Replaces a matching existing Job.
+   *
+   * @param jobName
+   * @param jobClassName
+   * @param params Set this null if there are no params
+   * @param isConcurrencyAllowed
+   */
+  public static void addJob(String jobName, String jobClassName, Map<String, Object> params, boolean isConcurrencyAllowed)
+      throws SundialSchedulerException {
+
+    try {
+      addJob(jobName, getScheduler().getCascadingClassLoadHelper().loadClass(jobClassName), params, isConcurrencyAllowed);
+    } catch (ClassNotFoundException e) {
+      throw new SundialSchedulerException("ERROR ADDING JOB!!!", e);
+    }
+  }
+
+  /**
+   * Starts a Job matching the given Job Name
+   *
+   * @param jobName
+   */
+  public static void startJob(String jobName) throws SundialSchedulerException {
+
+    try {
+      getScheduler().triggerJob(jobName, null);
+    } catch (SchedulerException e) {
+      throw new SundialSchedulerException("ERROR STARTING JOB!!!", e);
+    }
+  }
+
+  /**
+   * Removes a Job matching the given Job Name
+   *
+   * @param jobName
+   */
+  public static void removeJob(String jobName) throws SundialSchedulerException {
+
+    try {
+      getScheduler().deleteJob(jobName);
+    } catch (SchedulerException e) {
+      throw new SundialSchedulerException("ERROR REMOVING JOB!!!", e);
+    }
+
+  }
+
+  /**
+   * Starts a Job matching the the given Job Name found in jobs.xml
+   *
+   * @param jobName
+   */
+  public static void startJob(String jobName, Map<String, Object> params) throws SundialSchedulerException {
+
+    try {
+
+      JobDataMap jobDataMap = new JobDataMap();
+      for (String key : params.keySet()) {
+        // logger.debug("key= " + key);
+        // logger.debug("value= " + pParams.get(key));
+        jobDataMap.put(key, params.get(key));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: Iterating over [map.keySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#keySet--) and calling [map.get(Key)](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#get-java.lang.Object-) to get the values within the loop is inefficient. 
**Fix**: Iterate over [map.entrySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#entrySet--) and access the values using [entry.getValue()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.Entry.html#getValue--) to avoid potential performance penalty due to lookup.